### PR TITLE
reduce cost of style calc+layout in --dialog-scrollgutter

### PR DESCRIPTION
**Pull request by keithamus**
_Authors: Please fill out this form carefully and completely._

_Reviewers: By approving this Pull Request you are approving the code change, as well as its deployment and mitigation plans._
_Please read this description carefully. If you feel there is anything unclear or missing, please ask for updates._

### What are you trying to accomplish?

In our mob pairing session called "Making Dotcom Fast" we ran into an issue where the Primer Dialog component is causing an expensive recalc operation on every page load; this is due to it setting the `--dialog-scrollgutter` css property, which is a property derived from the `innerWidth` & `body.clientWidth` properties.

In order to prevent this cost on page load, we decided to move the operation to the first `click` event that can be handled by the `dialogInvokerButtonHandler`. This moves the cost of the recalc away from document load to document click:

image

image

However this change alone still leaves the recalc cost during the first click event which is still costly and could impact INP. 

image

So we added some CSS - inspired by [this article on the `@property` rule and how it can lower recalc costs by disabling inheritance](https://web.dev/blog/at-property-performance) - which is perfect for this use case. This reduced recalc cost from ~70-90ms on my machine to under 2ms:

image

We've _not_ moved the code away from the hot-path, and into the click event listener, due to compat risk with the Dialog element and its use in dotcom. This can come at a later date, perhaps.

### Screenshots

(See above perf traces)

#### Risk Assessment
  

- [x] **Low risk** the change is small, highly observable, and easily rolled back.
- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.

### Accessibility

- **Fixes axe scan violation** - This change fixes an existing [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation.
- **No new axe scan violation** - This change does not introduce any new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violations.
- **New axe violation** - This change introduces a new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation. Please describe why the violation cannot be resolved below.

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Added/updated previews (Lookbook)
- [x] Tested in Chrome
- [ ] Tested in Firefox
- [ ] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

/cc the commit co-authors:
@ansballard, @AdamShwert, @arelia, @arielvalentin, @chrisgavin, @Cbodfield, @composerinteralia, @garman, @dgreif, @erinnachen, @georgebrock, @jibrang, @jfuchs, @khiga8, @manuelpuyol, @mattcosta7, @another-mattr, @MelissaPastore, @theinterned, @phillmv, @Raffo, @cheshire137, @srt32

### CODE REVIEW

File: app/components/primer/dialog_helper.ts
Comment (dgreif) [inline app/components/primer/dialog_helper.ts:1]: Are there ever cases where a dialog can be open on page load? Or is a click _absolutely required_ to get a dialog open?

**Review by langermank — APPROVED** [review]
This is cool! 😄

File: app/components/primer/dialog_helper.ts
Comment (cheshire137) [reply app/components/primer/dialog_helper.ts:1]: > Are there ever cases where a dialog can be open on page load?

I know we do some of these internally, but the approach there is to use a [Catalyst component](https://github.com/github/catalyst) that, when connected, either calls `showModal` on an `HTMLDialogElement` or sets the `open` property for other elements. So I think that would still work with this change.

File: app/components/primer/alpha/dialog.pcss [hunk 1]
```diff
@@ -8,8 +8,15 @@
 
 /* Overlay */
 
+@property --dialog-scrollgutter {
```
Comment (mattcosta7) [inline app/components/primer/alpha/dialog.pcss:11]: nit: can we comment here, maybe with a link to this issue?  future us might wonder why we're doing it

File: app/components/primer/dialog_helper.ts
Comment (keithamus) [reply app/components/primer/dialog_helper.ts:1]: This does make it more challenging. Perhaps we can just ship the CSS fix for now. Or we can dispatch our own toggle events which that Catalyst component can emulate.

Comment (keithamus) [reply app/components/primer/dialog_helper.ts:1]: Given the compat risk involved here with this catalyst component, let's drop the JS for now and think about how to resolve the JS issue at a later time (maybe when events are added).

**Pull request merged**